### PR TITLE
SCC admission: ignore users and groups fields

### PR DESCRIPTION
**Pull request by stlaz**
Rely solely on RBAC to decide whether users can 'use' SCCs to
create their pods.

This also moves the bootstrapping obligations from SCC bootstrap
modules to those that handle roles/clusterroles.

/cc @openshift/sig-auth @enj

### CODE REVIEW

File: pkg/cmd/openshift-apiserver/openshiftapiserver/openshift_apiserver.go [hunk 1]
```diff
@@ -590,9 +590,6 @@ func (c *completedConfig) startProjectAuthorizationCache(context genericapiserve
 }
 
 func (c *completedConfig) bootstrapSCC(context genericapiserver.PostStartHookContext) error {
```
Comment (enj) [inline pkg/cmd/openshift-apiserver/openshiftapiserver/openshift_apiserver.go]: General comment: we need to kill the add/remove-scc-to-* commands

File: pkg/cmd/server/bootstrappolicy/constants.go [hunk 2]
```diff
@@ -115,6 +116,7 @@ const (
 	ClusterAdminRoleBindingName       = ClusterAdminRoleName + "s"
 	ClusterReaderRoleBindingName      = ClusterReaderRoleName + "s"
 	BasicUserRoleBindingName          = BasicUserRoleName + "s"
+	BasicSCCUserRoleBindingName       = BasicSCCUserRoleName + "s"
```
Comment (enj) [inline pkg/cmd/server/bootstrappolicy/constants.go]: no adding "s" to end - that was mistake

File: pkg/cmd/server/bootstrappolicy/controller_policy.go [hunk 3]
```diff
@@ -112,6 +112,7 @@ func init() {
 			rbacv1helpers.NewRule("get", "list").Groups(kapiGroup).Resources("serviceaccounts").RuleOrDie(),
 			rbacv1helpers.NewRule("create").Groups(securityGroup, legacySecurityGroup).Resources("podsecuritypolicysubjectreviews").RuleOrDie(),
 			rbacv1helpers.NewRule("get", "list").Groups(configGroup).Resources("builds").RuleOrDie(),
+			rbacv1helpers.NewRule("use").Groups(securityGroup).Resources("securitycontextconstraints").Names("privileged").RuleOrDie(),
```
Comment (enj) [inline pkg/cmd/server/bootstrappolicy/controller_policy.go]: does it still need this permission?

File: pkg/cmd/server/bootstrappolicy/controller_policy.go [hunk 4]
```diff
@@ -334,6 +335,7 @@ func init() {
 			rbacv1helpers.NewRule("get", "update", "list", "watch").Groups(kapiGroup).Resources("persistentvolumeclaims").RuleOrDie(),
 			rbacv1helpers.NewRule("update").Groups(kapiGroup).Resources("persistentvolumeclaims/status").RuleOrDie(),
 			rbacv1helpers.NewRule("get", "create", "delete", "list", "watch").Groups(kapiGroup).Resources("pods").RuleOrDie(),
+			rbacv1helpers.NewRule("use").Groups(securityGroup).Resources("securitycontextconstraints").Names("hostmount-anyuid").RuleOrDie(),
```
Comment (enj) [inline pkg/cmd/server/bootstrappolicy/controller_policy.go]: same question

File: pkg/cmd/server/bootstrappolicy/policy.go [hunk 5]
```diff
@@ -768,6 +779,9 @@ func GetOpenshiftBootstrapClusterRoleBindings() []rbacv1.ClusterRoleBinding {
 		newOriginClusterBinding(BasicUserRoleBindingName, BasicUserRoleName).
 			Groups(AuthenticatedGroup).
 			BindingOrDie(),
+		newOriginClusterBinding(BasicSCCUserRoleBindingName, BasicSCCUserRoleName).
```
Comment (enj) [inline pkg/cmd/server/bootstrappolicy/policy.go]: use `rbacv1helpers.NewClusterBinding`

File: pkg/cmd/server/bootstrappolicy/securitycontextconstraints.go [hunk 6]
```diff
@@ -296,30 +295,6 @@ func GetBootstrapSecurityContextConstraints(sccNameToAdditionalGroups map[string
...
-		}
-		if groupsToAdd, ok := sccNameToAdditionalGroups[constraint.Name]; ok {
-			constraints[i].Groups = append(constraints[i].Groups, groupsToAdd...)
-		}
 	}
 	return constraints
 }
-
-// GetBoostrapSCCAccess provides the default set of access that should be passed to GetBootstrapSecurityContextConstraints.
-func GetBoostrapSCCAccess(infraNamespace string) (map[string][]string, map[string][]string) {
```
Comment (enj) [inline pkg/cmd/server/bootstrappolicy/securitycontextconstraints.go]: General question: what of this still needs this access.  If it still needs this access, can the access be namespace scoped and/or managed by CVO manifest (i.e. moved out of origin).

File: pkg/cmd/server/bootstrappolicy/constants.go [hunk 7]
```diff
@@ -65,6 +65,7 @@ const (
 	AggregatedClusterReaderRoleName = "system:openshift:aggregate-to-cluster-reader"
 	SelfProvisionerRoleName         = "self-provisioner"
 	BasicUserRoleName               = "basic-user"
+	BasicSCCUserRoleName            = "basic-scc-user"
```
Comment (enj) [inline pkg/cmd/server/bootstrappolicy/constants.go]: Will need to update the bootstrap policy test YAML:

`UPDATE_BOOTSTRAP_POLICY_FIXTURE_DATA=true hack/test-go.sh pkg/cmd/server/bootstrappolicy`

File: pkg/oc/cli/admin/policy/reconcile_sccs.go [hunk 8]
```diff
@@ -211,8 +211,7 @@ func (o *ReconcileSCCOptions) ChangedSCCs() (
 	toUpdateSCCs := []*securityv1.SecurityContextConstraints{}
 	toCreateSCCs := []*securityv1.SecurityContextConstraints{}
 
-	groups, users := bootstrappolicy.GetBoostrapSCCAccess(o.InfraNamespace)
-	bootstrapSCCs := bootstrappolicy.GetBootstrapSecurityContextConstraints(groups, users)
+	bootstrapSCCs := bootstrappolicy.GetBootstrapSecurityContextConstraints()
```
Comment (enj) [inline pkg/oc/cli/admin/policy/reconcile_sccs.go]: the `if o.Union {` logic will need to be removed

File: pkg/security/apiserver/securitycontextconstraints/matcher.go [hunk 9]
```diff
@@ -99,21 +99,11 @@ func authorizedForSCC(sccName string, info user.Info, namespace string, a author
 // ConstraintAppliesTo inspects the constraint's users and groups against the userInfo to determine
```
Comment (enj) [inline pkg/security/apiserver/securitycontextconstraints/matcher.go]: out of date comment

**Comment by stlaz** [comment]
Added WIP tag, I'm working on fixing/working around the tests, I'll remove the roles for pv-recycle- and build- controllers, and, of course, need to remove the commands.

**Comment by stlaz** [comment]
Our integration test suite relies on the `pv-controller-recycler` and `build-controller` roles to be able to `use` some of the SCCs it would seem, adding them back for now, also the `monitoring` operator was failing, let's see if by adding those roles back, they start working again.

**Comment by stlaz** [comment]
/test integration e2e-aws-builds

Looks like openshift-apiserver died during `e2e-aws-builds`

**Comment by stlaz** [comment]
Console deployment did not come up:
/test e2e-aws-builds
openshift-apiserver (?) died:
/test e2e-aws-image-registry

**Comment by stlaz** [comment]
There may still be some gotchas in the tests as to "this should have been removed instead of trying to make it work", but this seems to work now.

File: pkg/oc/cli/admin/policy/reconcile_sccs.go (hunk 8 above)
Comment (stlaz) [reply pkg/oc/cli/admin/policy/reconcile_sccs.go]: I kept those parts that are irrelevant to this change

**Comment by stlaz** [comment]
I kept those permissions, at least one of them is required for the integration tests.

We should however do a bootstrap RBAC audit, I suspect some of the bootstrap roles are no longer used.

**Comment by ericavonb** [comment]
/hold
We didn't get an exception for this one - can come back to it after all the bugs are fixes 🙂

**Pull request closed without merging**